src/librustc/ty/mod.rs

```diff
@@ -742,14 +742,36 @@ impl Generics {
         self.parent_count() + self.own_count()
     }
 
-    pub fn region_param(&self, param: &EarlyBoundRegion) -> &RegionParameterDef {
-        assert_eq!(self.parent_count(), 0);
-        &self.regions[param.index as usize - self.has_self as usize]
+    pub fn region_param(&'tcx self,
+                        param: &EarlyBoundRegion,
+                        tcx: TyCtxt<'a, 'gcx, 'tcx>)
+                        -> &'tcx RegionParameterDef
+    {
+        if let Some(index) = param.index.checked_sub(self.parent_count() as u32) {
+            &self.regions[index as usize - self.has_self as usize]
+        } else {
+            tcx.generics_of(self.parent.expect("parent_count>0 but no parent?"))
+                .region_param(param, tcx)
+        }
     }
 
-    pub fn type_param(&self, param: &ParamTy) -> &TypeParameterDef {
-        assert_eq!(self.parent_count(), 0);
-        &self.types[param.idx as usize - self.has_self as usize - self.regions.len()]
+    /// Returns the `TypeParameterDef` associated with this `ParamTy`, or `None`
+    /// if `param` is `self`.
+    pub fn type_param(&'tcx self,
+                      param: &ParamTy,
+                      tcx: TyCtxt<'a, 'gcx, 'tcx>)
+                      -> Option<&TypeParameterDef> {
+        if let Some(idx) = param.idx.checked_sub(self.parent_count() as u32) {
+            let type_param_start = (self.has_self as usize) + self.regions.len();
+            if let Some(idx) = (idx as usize).checked_sub(type_param_start) {
+                Some(&self.types[idx])
+            } else {
+                None
```

Comment:
Why not `assert_eq!((idx, self.has_self), (0, true));` followed by `&self.types[0]`?

Reply:
`idx` of 1 is also mapped to `1 - 1 = 0`, so I don't think `&self.types[0]` is correct here and `Self` doesn't have a `TypeParameterDef`.